source/common/http/http3_status_tracker.cc

```diff
@@ -0,0 +1,46 @@
+#include "common/http/http3_status_tracker.h"
+
+namespace Envoy {
+namespace Http {
+
+namespace {
+
+// Initially, HTTP/3 is be marked broken for 5 minutes.
+const std::chrono::minutes DefaultExpirationTime{5};
+
+} // namespace
+
+Http3StatusTracker::Http3StatusTracker(Event::Dispatcher& dispatcher)
+    : expiration_timer_(dispatcher.createTimer([this]() -> void { onExpirationTimeout(); })) {}
+
+bool Http3StatusTracker::isHttp3Broken() const { return state_ == State::Broken; }
+
+bool Http3StatusTracker::isHttp3Confirmed() const { return state_ == State::Confirmed; }
+
+void Http3StatusTracker::markHttp3Broken() {
+  state_ = State::Broken;
```

Comment:
What are valid values for state_ when entering this method?

Reply:
It may be possible to trigger this ASSERT, if there are 2 concurrent attempts to connect to the same endpoint.  That said, it's possible that there are protections elsewhere to prevent this from happening or it is relatively unlikely to happen without a burst of requests for that service; we would need the number of requests to exceed the upstream's multiplexing factor and trigger creation of a second connection in order to meet demand.

Reply:
Just for my own curiosity why would we have multiple concurrent connection attempts to the same endpoint on a given worker thread? Is this related to prefetching or part of the grid logic?

Reply:
I'm relatively unclear at this on exactly how this all plays together. From what Antonio said, it sound like if there were a sufficient number of simultaneous requests we might trigger the creation of a second attempt. The other case that I *think* I heard from alyssa is that it's possible to have multiple calls to ConnectivityGrid::newStream() happen before the first call finishes. This won't result in multiple TCP/QUIC connection attempts because the underlying connection pool will do the right thing. But I *think* this is transparent to the ConnectivtyGrid; each call to newStream creates a new WrapperCallbacks and (up to) 2 ConnectionAttempts which should mean it's possible to get unexpected state transitions. Happy to do something different if I'm not understanding.

Reply:
Yeah this sounds plausible to me: newStream is non-blocking (like most things in Envoy), so if multiple streams are established before the connection is established so you'd see multiple newStreams come in before the connection is established.

No change necessary, I was just curious how this all fit together :)

Reply:
Optional (I mean in this PR, not optional overall) I wonder if we should start landing docs on this as we start implementing the "real" logic.

Generally we land docs when we unhide config (https://github.com/envoyproxy/envoy/pull/15926) but the failover logic is sufficiently complicated I think we could land docs for now in source/docs and then move them to docs/ when the PR lands.  Your call if we do them now or in a future iteration :-)

Reply:
Docs definitely make sense and I'm happy to work on them. I think I'll do that in a follow-up since this PR is (hopefully) basically done at this point.